File: frontend/src/types/kanban.ts
import type { ProjectPriority } from "@/types/project";

export interface KanbanColumn {
  id: string;
  project_id: string;
  name: string;
  column_type: "todo" | "in_progress" | "done" | "custom";
  position: number;
  color?: string | null;
  created_at: string;
}

export interface KanbanTaskField {
  id: string;
  task_id: string;
  name: string;
  value: string;
  position: number;
  created_at: string;
  updated_at: string;
}

export interface TaskFieldDraft {
  name: string;
  value: string;
}

// Custom task-field limits. Keep in sync with the backend validation in
// backend/internal/dto/operational/kanban_task_field.go.
export const MAX_TASK_FIELDS = 50;
export const MAX_TASK_FIELD_NAME = 120;
export const MAX_TASK_FIELD_VALUE = 20000;

export interface KanbanTaskListItem {
  id: string;
  column_id: string;
  title: string;
  description?: string | null;
  assignee_id?: string | null;
  assignee_name?: string | null;
  avatar_url?: string | null;
  due_date?: string | null;
  priority: ProjectPriority;
  label?: string | null;
  assigned_via: "manual" | "auto";
  position: number;
}

export interface KanbanTask {
  id: string;
  column_id: string;
  project_id: string;
  title: string;
  description?: string | null;
  assignee_id?: string | null;
  assignee_name?: string | null;
  avatar_url?: string | null;
  due_date?: string | null;
  priority: ProjectPriority;
  label?: string | null;
  assigned_via: "manual" | "auto";
  position: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  fields?: KanbanTaskField[] | null;
}

export interface KanbanFilters {
  assignee: string;
  priority: string;
  label: string;
  dueDate: string;
  search: string;
  columnId: string;
}

export interface TaskFormValues {
  title: string;
  description: string;
  assignee_id: string;
  due_date: string;
  priority: ProjectPriority;
  label: string;
  fields: TaskFieldDraft[];
}

File: backend/internal/model/kanban.go
package model

import "time"

const (
	KanbanColumnTypeTodo       = "todo"
	KanbanColumnTypeInProgress = "in_progress"
	KanbanColumnTypeDone       = "done"
	KanbanColumnTypeCustom     = "custom"

	KanbanTaskAssignedViaManual = "manual"
	KanbanTaskAssignedViaAuto   = "auto"
)

type KanbanColumn struct {
	ID         string    `json:"id"`
	ProjectID  string    `json:"project_id"`
	Name       string    `json:"name"`
	ColumnType string    `json:"column_type"`
	Position   int       `json:"position"`
	Color      *string   `json:"color,omitempty"`
	CreatedAt  time.Time `json:"created_at"`
}

type KanbanTaskListItem struct {
	ID           string     `json:"id"`
	ColumnID     string     `json:"column_id"`
	Title        string     `json:"title"`
	Description  *string    `json:"description,omitempty"`
	AssigneeID   *string    `json:"assignee_id,omitempty"`
	AssigneeName *string    `json:"assignee_name,omitempty"`
	AvatarURL    *string    `json:"avatar_url,omitempty"`
	DueDate      *time.Time `json:"due_date,omitempty"`
	Priority     string     `json:"priority"`
	Label        *string    `json:"label,omitempty"`
	AssignedVia  string     `json:"assigned_via"`
	Position     int        `json:"position"`
}

func (t KanbanTask) ToListItem() KanbanTaskListItem {
	return KanbanTaskListItem{
		ID:           t.ID,
		ColumnID:     t.ColumnID,
		Title:        t.Title,
		Description:  t.Description,
		AssigneeID:   t.AssigneeID,
		AssigneeName: t.AssigneeName,
		AvatarURL:    t.AvatarURL,
		DueDate:      t.DueDate,
		Priority:     t.Priority,
		Label:        t.Label,
		AssignedVia:  t.AssignedVia,
		Position:     t.Position,
	}
}

type KanbanTask struct {
	ID           string     `json:"id"`
	ColumnID     string     `json:"column_id"`
	ProjectID    string     `json:"project_id"`
	Title        string     `json:"title"`
	Description  *string    `json:"description,omitempty"`
	AssigneeID   *string    `json:"assignee_id,omitempty"`
	AssigneeName *string    `json:"assignee_name,omitempty"`
	AvatarURL    *string    `json:"avatar_url,omitempty"`
	DueDate      *time.Time `json:"due_date,omitempty"`
	Priority     string     `json:"priority"`
	Label        *string    `json:"label,omitempty"`
	AssignedVia  string     `json:"assigned_via"`
	Position     int        `json:"position"`
	CreatedBy    string     `json:"created_by"`
	CreatedAt    time.Time         `json:"created_at"`
	UpdatedAt    time.Time         `json:"updated_at"`
	Fields       []KanbanTaskField `json:"fields"`
}
